### PR TITLE
add align_denoise implementation

### PR DESCRIPTION
**Pull request by bobchennan**
Non-autoregressive model: Align-Denoise

### CODE REVIEW

File: espnet/nets/pytorch_backend/ctc.py [hunk 1]
```diff
@@ -80,6 +83,31 @@ def loss_fn(self, th_pred, th_target, th_ilen, th_olen):
...
             raise NotImplementedError
 
+    def ctc_alignment_targets(self, hs_pad, hlens, ys_pad):
+        with torch.no_grad():
+            ys = [y[y != self.ignore_id] for y in ys_pad]  # parse padded ys
+    
+            hlens = torch.from_numpy(np.fromiter(hlens, dtype=np.int64))
+            olens = torch.from_numpy(np.fromiter(
+                (x.size(0) for x in ys), dtype=np.int64))
+            ys_hat = self.ctc_lo(F.dropout(hs_pad, p=self.dropout_rate)).transpose(0, 1).log_softmax(dim=-1)
```
Comment (pengchengguo) [inline espnet/nets/pytorch_backend/ctc.py:93]: `F.dropout()` needs to have an optional parameter named `self.training` to distinguish the training and inference phase. Does the `ctc_alignment_targets` function only be used in the training phase?
Related PR: https://github.com/espnet/espnet/pull/3539

File: espnet/nets/pytorch_backend/e2e_asr_aligndenoise.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        
+        batch_size = xs_pad.size(0)
+        hs_len = hs_mask.view(batch_size, -1).sum(1)
+        loss_ctc = self.ctc(hs_pad.view(batch_size, -1, self.adim), hs_len, ys_pad)
+        ys_hat = self.ctc.argmax(hs_pad.view(batch_size, -1, self.adim)).data
+        cer_ctc = self.error_calculator(ys_hat.cpu(), ys_pad.cpu(), is_ctc=True)
+
+        
+        if not self.training:
+            self.ctc.softmax(hs_pad)
```
Comment (pengchengguo) [inline espnet/nets/pytorch_backend/e2e_asr_aligndenoise.py:120]: No tensor is assigned to the return value.

**Comment by pengchengguo** [comment]
Could you please please fix the black format error, thanks!

File: espnet/nets/pytorch_backend/ctc.py (hunk 1 above)
Comment (bobchennan) [reply espnet/nets/pytorch_backend/ctc.py:93]: This is only used in the training to get the force alignment from CTC.

Comment (bobchennan) [reply espnet/nets/pytorch_backend/ctc.py:93]: But you are right this is something interesting to consider

File: espnet/nets/pytorch_backend/e2e_asr_aligndenoise.py (hunk 2 above)
Comment (bobchennan) [reply espnet/nets/pytorch_backend/e2e_asr_aligndenoise.py:120]: https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/e2e_asr_transformer.py#L226

It is for visualization. I don't understand the design either lol.
